### PR TITLE
Initial draft for implementing full width responsive non fast-fetch path

### PR DESCRIPTION
**Pull request by riklund**
Initial draft for implementing full width responsive non fast-fetch path.

This PR doesn't contain any tests yet as we're interested to hear thoughts about this approach before adding them.

### CODE REVIEW

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 1]
```diff
@@ -300,4 +376,32 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+  attemptFullWidthSizeChange_() {
+    const viewportSize = this.getViewport().getSize();
+    const minHeight = 100;
+    const maxHeight = Math.min(500, viewportSize.height);
+    const ratio = this.config.fullWidthHeightRatio;
+    dev().assert(!!ratio);
+    const idealHeight = Math.round(viewportSize.width / ratio);
+    const height = clamp(idealHeight, minHeight, maxHeight);
+    const width = viewportSize.width;
+    // Attempt to resize to the correct height. The width should already be
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: To confirm
> width should already be 100vw, 

is this required? For example would 
```
<amp-ad width="50vw" height=150 data-full-width>
```
be valid?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 2]
```diff
@@ -193,6 +249,26 @@ export class AmpAd3PImpl extends AMP.BaseElement {
     if (this.xOriginIframeHandler_) {
       this.xOriginIframeHandler_.onLayoutMeasure();
     }
+
+    if (this.fullWidthState_ == FullWidthState.REQUESTED) {
+      this.fullWidthState_ = FullWidthState.EXPANDED;
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: We are setting state to EXPANDED even when attemptChangeSize fail?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 3]
```diff
@@ -193,6 +249,26 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+
+      // Nudge into the correct horizontal position by changing side margin.
+      this.getVsync().run({
+        measure: state => {
+          state.direction =
+              computedStyle(this.win, this.element)['direction'];
+        },
+        mutate: state => {
+          if (state.direction == 'rtl') {
+            setStyle(this.element, 'marginRight', layoutBox.left, 'px');
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js:257]: Are we not respecting the style setting from body here?

File: ads/_config.js [hunk 4]
```diff
@@ -118,6 +124,7 @@ export const adConfig = {
     clientIdCookieName: '_ga',
     remoteHTMLDisabled: true,
     masterFrameAccessibleType: 'google_network',
+    fullWidthHeightRatio: 1.2
```
Comment (zhouyx) [inline ads/_config.js]: Question: the ratio is per ad network defined not per ad?

Comment (riklund) [reply ads/_config.js]: Correct - an ad network defines a ratio and it'll be the same for all ads on that network.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 2 above)
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Correct - we still do the margin adjustments and we only want to do those exactly once.
(If attemptChangeSize fails the height will be unchanged, but that should be fine as it should be a minority of cases)

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 3 above)
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js:257]: Sorry, not sure I understand - could you clarify the question?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 1 above)
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: No, we'll require 100vw.

**Review by riklund — COMMENTED** [review]
Thanks!

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 3 above)
Comment (zhouyx) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js:257]: Say publisher set body `padding: 10px`, are we overriding that?

Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js:257]: Yes, in the sense that we'll align the ad unit to the edge of the screen anyway.
(that's by design)

**Pull request closed without merging**